Fix media uploads error handling

kean opened the pull request:
Part of https://github.com/wordpress-mobile/WordPress-iOS/issues/21190. 

Fixes missing video upload limits reported by @momo-ozawa [here](https://github.com/wordpress-mobile/WordPress-iOS/pull/21398#pullrequestreview-1593143993). The fix requires a lot of changes because I found multiple other production issues on the Media screen that had to be fixed.

@osullivanchris, if you have a better idea of how to handle this scenario, given the technical limitations, please let me know. It may also be worth re-visiting this entire flow in the scope of the Media changes.

## Handling Upload Limits

One of the advantages of using a custom picker is that it allows us to support scenarios like the upload limits. The custom picker displays the errors in the picker itself ([recording.mp4](https://github.com/wordpress-mobile/WordPress-iOS/assets/1567433/55cbcd84-3707-49a9-b57c-9732e7995fb5)).

<img width="320" alt="current-screenshot" src="https://github.com/wordpress-mobile/WordPress-iOS/assets/1567433/549bbfe2-6780-4c7b-9fcb-338b9e8bfee4">

Unfortunately, it wasn't possible with `PHPickerViewController`. I implemented it by handling this as other export errors ([recording.mp4](https://github.com/wordpress-mobile/WordPress-iOS/assets/1567433/e543616d-4775-47c1-b636-e31d59a8ad3c))

<img width="320" alt="Screenshot 2023-08-24 at 4 37 39 PM" src="https://github.com/wordpress-mobile/WordPress-iOS/assets/1567433/03d5720d-5ef5-4416-9a1a-86cd09424148">  <img width="320" alt="Screenshot 2023-08-24 at 4 53 39 PM" src="https://github.com/wordpress-mobile/WordPress-iOS/assets/1567433/f050e93f-524a-48e9-b731-2d15f6f74c37">


To implement it, I added duration limit checks to `MediaVideoExporter`. It was supposed to just work, but I encountered a few issues I had to fix.

### 1. Fix slow exports from NSItemProvider (https://github.com/wordpress-mobile/WordPress-iOS/commit/0376ca171faae2b0aa46a814b9938e2f894b02c7)

The `NSItemProvider` returned by `PHPickerViewController` doesn't return the video duration – you have to load the file first. But `loadFileRepresentation` was too slow: it was taking more than a minute for my 500 MB video. The reason was transcoding.

I followed the [advice](https://developer.apple.com/forums/thread/652496) of an Apple engineer to disable transcoding on `PHPickerViewController`. The `loadFileRepresentation` was now taking around 200 ms. It’s the correct course of action because [MediaVideoExporter.swift](https://github.com/wordpress-mobile/WordPress-iOS/blob/trunk/WordPress/Classes/Utility/Media/MediaVideoExporter.swift) also performs its own transcoding. We definitely don’t want to do it twice. 

### 2. Fix an issue with MediaImportService not reporting errors (https://github.com/wordpress-mobile/WordPress-iOS/commit/13285a395743dd3aaec2bc9a12bf03199e9a5286)

`MediaImportService` was incorrectly calling a completion closure with `.success` on export failures. The errors were never reported to the Media screen.

### 3. Fix invalid localizedDescription of exporter errors (https://github.com/wordpress-mobile/WordPress-iOS/commit/c832575dbfed8b8ce4d1a5bf1f9aa8f141795020)

Before -> After

<img width="320" alt="Screenshot 2023-08-24 at 3 23 45 PM" src="https://github.com/wordpress-mobile/WordPress-iOS/assets/1567433/be4f77db-9496-4bd0-a719-b3f6140a039b"> <img width="320" alt="Screenshot 2023-08-24 at 4 37 39 PM" src="https://github.com/wordpress-mobile/WordPress-iOS/assets/1567433/a9b57f3b-47ec-4997-a107-2f05925e52ef">

### 4. Fix incorrect error banner displayed when upload fails https://github.com/wordpress-mobile/WordPress-iOS/commit/c144f66e4bd6cb250c0af1c523fb19db44c57d0e

Before -> After

<img width="320" alt="Screenshot 2023-08-24 at 3 42 15 PM" src="https://github.com/wordpress-mobile/WordPress-iOS/assets/1567433/d25253c5-54da-45a1-b12f-7c237b732e58"> <img width="320" alt="Screenshot 2023-08-24 at 3 47 59 PM" src="https://github.com/wordpress-mobile/WordPress-iOS/assets/1567433/29e8eb4c-84d9-45e8-bd75-32214f38f1e3">

## To test:

- Add a long video (5 minutes to Photos)
- Select a site with a free plan
- Open Media and try uploading the video
- Verify that the error is shown

Ideally we also need to re-run the tests from https://github.com/wordpress-mobile/WordPress-iOS/pull/21343

## Regression Notes

1. Potential unintended areas of impact: Media
2. What I did to test those areas of impact (or what existing automated tests I relied on): manual
3. What automated tests I added (or what prevented me from doing so): n/a

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

UI Changes testing checklist:
- [ ] Portrait and landscape orientations.
- [ ] Light and dark modes.
- [ ] Fonts: Larger, smaller and bold text.
- [ ] High contrast.
- [ ] VoiceOver.
- [ ] Languages with large words or with letters/accents not frequently used in English.
- [ ] Right-to-left languages. (Even if translation isn’t complete, formatting should still respect the right-to-left layout)
- [ ] iPhone and iPad. 
- [ ] Multi-tasking: Split view and Slide over. (iPad)

kean committed:
Improve logging in ItemProviderMediaExporter

kean committed:
Switch to .current representation mode to prevent slow transcoding

The issue wasn't just that the transcoding was slow, but that the app also performs transcoding in MediaVideoExporter, duplicating the effort. One more advantage of removing the transcoding is that we can now quickly check the video duration before the upload.

kean committed:
Fix an issue with MediaCoordinator not reporting import errors

kean committed:
Fix invalid localizedDescription of exporter errors

kean committed:
Fix an issue with incorrect error banner displayed when media fails to upload on the Media screen

kean committed:
Refactor Blog.canUploadAsset

kean committed:
Add video upload limits

kean commented on WordPress/WordPressTest/BlogVideoLimitsTests.swift:
```diff
@@ -1,56 +0,0 @@
-        blog.hasPaidPlan = false
-
-        // Then it can upload assets when allowance not exceeded
-        XCTAssertTrue(blog.canUploadAsset(false))
```
It was basically testing configuration, repeating the logic one-to-one – not very useful.
Since I removed the `blog.canUploadAsset` variant the test was using, I decided to remove the test as well.
I'm planning to add some more tests for Media features as I start working on the screen itself.

momo-ozawa reviewed (commented):
- When I add a long video via the Media screen, then tap "retry" on the notice, the video is automatically deleted. However, tapping "retry" on the media thumbnail brings up the action sheet. Wdyt of showing the action sheet when tapping "retry" on the notice for consistency?
- On the action sheet, "cancel upload" and "delete" both seem to delete the video. Should we show only one option, if they both do the same thing?
- When adding a long video via the video block, you get a "failed to insert media" message. If you tap on it, an action sheet comes up. If you tap "retry", nothing seems to happen - and the action sheet doesn't come up again when you tap the video block.

osullivanchris commented:
It feels like we are really bending the limits of what these components are intended for. Its a bit of a strange UX. I mean, already, not just what we are adding.

My first idea was that for the upload limit, its not really an error. Its more of a flow/feature. We could be more proactive and show something like a sheet which informs the user about what happens. Potentially even with an upsell, if they are willing to subscribe to upload larger videos. 
![IMG_2852 2](https://github.com/wordpress-mobile/WordPress-iOS/assets/50576199/485dacbf-237b-4dfa-80c6-f4a1b80a3daf)
e1)

The second idea I had was that we are really limiting ourselves by trying to show an error within the grid. What if we didn't show an image within the tiled grid until it finished uploading. Provides us a space to show actual information about an unsuccessful upload and what actions to take, rather than having to squeeze info into snackbar messages, on top of the tiles, or in the action sheet heading.

![IMG_2854](https://github.com/wordpress-mobile/WordPress-iOS/assets/50576199/74387bc8-cdb7-4dcf-8f61-c9a6b8f3d038)


Maybe these ideas are too large. I don't think you're making things worse with the approach you've taken. It's following the current convention. So I'm ok if we just want to do that. Maybe only show one negative action as @momo-ozawa mentioned if they do the same thing.

kean commented:
> Wdyt of showing the action sheet when tapping "retry" on the notice for consistency?

Good catch with the notice view. I didn't check what it does.
There is no "Retry" at this point. The only thing you can do is delete the media item. I'll see how I can fix it.

> My first idea was that for the upload limit, its not really an error. 

Absolutely. Ideally, it should give you a way to upgrade and automatically resume the upload. But it will require some significant work.

> Maybe these ideas are too large. 

I'd suggest considering these changes in the scope of the other Media work. This PR's goal is to address the immediate issue of bringing the native picker integration to a state where we can ship it.

Again, we could take some inspiration from some of the native apps.

kean committed:
Remove duplicated Delete action

kean committed:
Stop showing Retry action for media uploads where retry is impossible

kean committed:
Hide the retry action when retry is not available

kean commented:
Thanks again for bringing up these issues @momo-ozawa.

I reviewed more of this code. There are two technical limitations/factors to consider:
- The notices are handled by the `MediaCoordinator`, which is global and has no connection to the media screen
- The "retry" action is only applicable when the export is successful (localURL != nil). `MediaCoordinator` will **not** try to re-export the asset, and in most/all cases there is no point to.

> Wdyt of showing the action sheet when tapping "retry" on the notice for consistency?

Given the technical limitations, I decided to not show the "Retry" button in this scenario ([screenshot](https://github.com/wordpress-mobile/WordPress-iOS/assets/1567433/bf3baf53-fdb5-444c-9ea2-c36f3c29caf1)).

> Should we show only one option, if they both do the same thing?

I removed the "Delete" action ([screenshot](https://github.com/wordpress-mobile/WordPress-iOS/assets/1567433/4e95ac9f-4921-4584-a282-552dfb5b6f8b)).

> the action sheet doesn't come up again when you tap the video block.

I removed the "Retry" action. It's not consistent with the Media screen, and tapping other actions seems to work the way you would expect ([screenshot](https://github.com/wordpress-mobile/WordPress-iOS/assets/1567433/73061a70-6a61-4f21-acb0-16cbb6ee6057)).

P.S. I also want to add Sentry error tracking for export errors to monitor it in the upcoming release, but that's going to be a separate PR.

momo-ozawa commented on WordPress/Classes/ViewRelated/Media/MediaProgressCoordinatorNoticeViewModel.swift:
```diff
@@ -59,16 +59,24 @@ struct MediaProgressCoordinatorNoticeViewModel {
     }
 
     private var failureNotice: Notice {
+        var canRetry = failedMedia.allSatisfy(\.canRetry)
```
TIL allSatisfy 👀

momo-ozawa reviewed (approved):
Verified
- "retry" button isn't shown in the notices
- "delete" action has been removed
- "retry" action has been removed for video block

LGTM! 🚀

kean committed:
Update release notes
